### PR TITLE
[dynamo] Fix cuda_stream pointer extraction for generic torch.Stream

### LINKED ISSUE #180179: torch.cuda.Stream traced incorrectly by torch.compile
### 🐛 Describe the bug

A fix to the previously reported bug , [#173172](https://github.com/pytorch/pytorch/issues/173172) appears to only work for the default stream. I believe the fix reads 'stream_id' into 'cuda_stream' pointer, from what I see in debug print. For default stream, both are 0 so that works:

root@aea89da2c313:/git/kernelcatcher/cuequivariance_ops_torch/tests# python compile_repro.py
current_cuda_stream: id = 0 cuda_stream = 0
In the op, cuda_stream = 0
no-stream call successful
Created stream <torch.cuda.Stream device=cuda:0 cuda_stream=0x1ee529d0>, id = 3 cuda_stream = 1ee529d0
current_cuda_stream: id = 3 cuda_stream = 1ee529d0
In the op, cuda_stream = 3
Segmentation fault (core dumped)

Please find repro script attached. 
You need to 'pip install cuda-bindings cuequivariance_ops_torch_cu13' (or cuequivariance_ops_torch_cu12 on CUDA12). 

[compile_repro.py](https://github.com/user-attachments/files/26661303/compile_repro.py)

### Versions

Collecting environment information...
PyTorch version: 2.12.0.dev20260411+cu130
Is debug build: False
CUDA used to build PyTorch: 13.0
ROCM used to build PyTorch: N/A

OS: Ubuntu 24.04.4 LTS (x86_64)
GCC version: (Ubuntu 13.3.0-6ubuntu2~24.04.1) 13.3.0
Clang version: Could not collect
CMake version: version 3.31.6
Libc version: glibc-2.39

Python version: 3.12.3 (main, Mar  3 2026, 12:15:18) [GCC 13.3.0] (64-bit runtime)
Python platform: Linux-6.8.0-79-generic-x86_64-with-glibc2.39
Is CUDA available: True
CUDA runtime version: 13.2.51
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: NVIDIA RTX A6000
Nvidia driver version: 580.82.07
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.9.20.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv.so.9.20.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn.so.9.20.0
/usr/lib/x86_64-linux-gnu/libcudnn_engines_precompiled.so.9.20.0
/usr/lib/x86_64-linux-gnu/libcudnn_engines_runtime_compiled.so.9.20.0
/usr/lib/x86_64-linux-gnu/libcudnn_graph.so.9.20.0
/usr/lib/x86_64-linux-gnu/libcudnn_heuristic.so.9.20.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops.so.9.20.0
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True
Caching allocator config: N/A

cc @ptrblck @msaroufim @eqy @jerryzh168 @tinglvv @nWEIdia @chauhang @penguinwu @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @kadeng @amjames @Lucaskabela @jataylo @azahed98

### PR DESCRIPTION
**Pull request by GSumanth109**
Fixes #180179 

### Summary
I noticed that `torch.compile` narrows `torch.cuda.Stream` down to generic `torch.Stream` objects during tracing. This is a problem because generic streams don't have the `.cuda_stream` property, so Dynamo falls back to returning the `stream_id` (a small int like `3`) instead of the actual hardware pointer. This leads to segfaults or "illegal memory access" errors when that integer is passed into a CUDA kernel expecting a real `cudaStream_t` address.

### Fix
I updated the logic in `CudaStreamVariable.var_getattr` to use a `hasattr` check instead of a strict type check, which lets us handle narrowed or proxied stream objects. I also added a recovery path that pulls the hardware pointer from the current CUDA context if the attribute is missing, while ensuring the default stream (ID 0) still returns a literal 0 to maintain parity with Eager mode and avoid breaking existing behavior.

### Testing
Added `test/dynamo/test_stream_pointer.py` to cover a few cases and verified that the compiled pointer matches the eager pointer for custom streams. Also verified whether the returned value is a real memory address (not just a truncated ID). It avoids issues from #173172 too.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @kadeng @chauhang @amjames @Lucaskabela @jataylo @azahed98 @mlazos

### CODE REVIEW

**Comment by mlazos** [comment]
@claude does this paper over any edge cases? I feel like it's a bug that `torch.Stream` doesn't support this or are we accessing a torch.cuda.stream field

File: torch/_dynamo/variables/streams.py [hunk 1]
```diff
@@ -549,11 +549,18 @@ def var_getattr(self, tx: "InstructionTranslator", name: str) -> VariableTracker
...
                 return ConstantVariable.create(self.value.cuda_stream)
+
             # For torch.Stream values (e.g. from torch.accelerator.current_stream),
             # the default stream has cuda_stream == stream_id == 0.
-            return ConstantVariable.create(self.value.stream_id)
+            if getattr(self.value, "stream_id", -1) == 0:
+                return ConstantVariable.create(0)
+
+            import torch.cuda
+            return ConstantVariable.create(torch.cuda.current_stream().cuda_stream)
```
Comment (mlazos) [inline torch/_dynamo/variables/streams.py]: this is incorrect, this could be any torch.cuda.stream instance not just the current stream (in theory), we need to extract the correct value from `self.value.native_handle`

File: test/dynamo/test_stream_pointer.py [hunk 2]
```diff
@@ -0,0 +1,22 @@
+import torch
```
Comment (mlazos) [inline test/dynamo/test_stream_pointer.py]: please don't add a new file, add this to test_streams.py

**Review by mlazos — CHANGES_REQUESTED** [review]

**Comment by GSumanth109** [comment]
Hey @mlazos, thanks for the catch on the current_stream() logic. I’ve refactored that to use native_handle instead, which is much safer and keeps the pointers tied to the specific stream instance as you suggested.

I also consolidated the tests into test/dynamo/test_streams.py and added a few extra scenarios to cover the edge cases @claude-bot flagged (like querying a stream that isn't currently active). Everything's passing on my end, and I've given the code a quick pass with the linter to keep it clean.

Let me know if any additional changes are to be made. Looking forward to your remarks!

**Comment by mlazos** [comment]
@GSumanth109 lint is still failing

File: torch/_dynamo/variables/streams.py [hunk 3]
```diff
@@ -549,13 +549,20 @@ def var_getattr(self, tx: "InstructionTranslator", name: str) -> VariableTracker
...
+            if hasattr(self.value, "cuda_stream"):
                 return ConstantVariable.create(self.value.cuda_stream)
+            if hasattr(self.value, "native_handle"):
+                return ConstantVariable.create(self.value.native_handle)
+
             # For torch.Stream values (e.g. from torch.accelerator.current_stream),
             # the default stream has cuda_stream == stream_id == 0.
-            return ConstantVariable.create(self.value.stream_id)
-        return super().var_getattr(tx, name)
+            if getattr(self.value, "stream_id", -1) == 0:
```
Comment (mlazos) [inline torch/_dynamo/variables/streams.py]: Remove this and the comment above

File: torch/_dynamo/variables/streams.py [hunk 4]
```diff
@@ -549,13 +549,20 @@ def var_getattr(self, tx: "InstructionTranslator", name: str) -> VariableTracker
 
             if self.source:
                 install_guard(self.source.make_guard(GuardBuilder.EQUALS_MATCH))
-            if isinstance(self.value, torch.cuda.Stream):
+
+            if hasattr(self.value, "cuda_stream"):
                 return ConstantVariable.create(self.value.cuda_stream)
+            if hasattr(self.value, "native_handle"):
```
Comment (mlazos) [inline torch/_dynamo/variables/streams.py]: we should just assert it hasattr, and return the native_handle result

**Comment by GSumanth109** [comment]
Hey @mlazos, I  wrapped the long lines to fix the `lintrunner` failures and verified them. Also updated the logic to assert `native_handle` and added the `@requires_cuda` decorator to the test as requested. This should resolve the trunk failures on the non-CUDA runners. Let me know if this looks good to go!

File: torch/_dynamo/variables/streams.py [hunk 5]
```diff
@@ -543,17 +543,23 @@ class CudaStreamVariable(StreamVariable):
...
                 install_guard(self.source.make_guard(GuardBuilder.EQUALS_MATCH))
-            if isinstance(self.value, torch.cuda.Stream):
+
+            if hasattr(self.value, "cuda_stream"):
                 return ConstantVariable.create(self.value.cuda_stream)
-            # For torch.Stream values (e.g. from torch.accelerator.current_stream),
-            # the default stream has cuda_stream == stream_id == 0.
-            return ConstantVariable.create(self.value.stream_id)
+
+            assert hasattr(self.value, "native_handle"), (
```
Comment (mlazos) [inline torch/_dynamo/variables/streams.py]: Instead of an assert can you do an if like above? it will fallthrough to unsupported at 563 so this is ok

**Review by mlazos — APPROVED** [review]
Please address my last comment, but I'd like to get this in asap

File: torch/_dynamo/variables/streams.py (hunk 5 above)
Comment (GSumanth109) [reply torch/_dynamo/variables/streams.py]: Just swapped it to an if. Let me know if its good to go!

**Review by mlazos — APPROVED** [review]
